scripts/automated_ingestion/eessi_task.py

```diff
@@ -0,0 +1,1201 @@
+
+import base64
+import os
+import subprocess
+import traceback
+
+from eessi_data_object import EESSIDataAndSignatureObject
+from eessi_logging import log_function_entry_exit, log_message, LoggingScope
+from eessi_task_action import EESSITaskAction
+from eessi_task_description import EESSITaskDescription
+from eessi_task_payload import EESSITaskPayload
+from utils import send_slack_message
+
+from github import Github, GithubException, InputGitTreeElement, UnknownObjectException
+from github.Branch import Branch
+from github.PullRequest import PullRequest
+
+
+@total_ordering
+class EESSITaskState(Enum):
+    UNDETERMINED = auto()  # The task state was not determined yet
+    NEW_TASK = auto()  # The task has been created but not yet processed
+    PAYLOAD_STAGED = auto()  # The task's payload has been staged to the Stratum-0
+    PULL_REQUEST = auto()  # A PR for the task has been created or updated in some staging repository
+    APPROVED = auto()  # The PR for the task has been approved
+    REJECTED = auto()  # The PR for the task has been rejected
+    INGESTED = auto()  # The task's payload has been applied to the target CernVM-FS repository
+    DONE = auto()  # The task has been completed
+
+    @classmethod
+    def from_string(
+        cls, name: str, default: Optional["EESSITaskState"] = None, case_sensitive: bool = False
+    ) -> "EESSITaskState":
+        log_message(LoggingScope.TASK_OPS_DETAILS, "INFO", "from_string: '%s'", name)
+        if case_sensitive:
+            to_return = cls.__members__.get(name, default)
+            log_message(LoggingScope.TASK_OPS_DETAILS, "INFO", "from_string will return: '%s'", to_return)
+            return to_return
+
+        try:
+            to_return = cls[name.upper()]
+            log_message(LoggingScope.TASK_OPS_DETAILS, "INFO", "from_string will return: '%s'", to_return)
+            return to_return
+        except KeyError:
+            return default
+
+    def __lt__(self, other):
+        if self.__class__ is other.__class__:
+            return self.value < other.value
+        return NotImplemented
+
+    def __str__(self):
+        return self.name.upper()
+
+
+class EESSITask:
+    description: EESSITaskDescription
+    payload: EESSITaskPayload
+    action: EESSITaskAction
+    git_repo: Github
+    config: Dict
+
+    @log_function_entry_exit()
+    def __init__(self, description: EESSITaskDescription, config: Dict, cvmfs_repo: str, git_repo: Github):
+        self.description = description
+        self.config = config
+        self.cvmfs_repo = cvmfs_repo
+        self.git_repo = git_repo
+        self.action = self._determine_task_action()
+
+        # define valid state transitions for all actions
+        # NOTE, for EESSITaskState.PULL_REQUEST, EESSITaskState.APPROVED must be the first element or
+        #   _next_state() will not work correctly
+        self.valid_transitions = {
+            EESSITaskState.UNDETERMINED: [
```

Comment:
I'm a bit confused. Here, you say essentially every state is a valid next state for `EESSITaskState.UNDETERMINED`. But then when I look at the handler for the undetermined task state, the only potential return values are UNDETERMINED (not part of this list!) or NEW_TASK?